Bluetooth: controller: Fix ticker collision update race

mtpr-ot opened the pull request:
In the case where a BLE connection ticker node collides and does not
trigger a callback to ULL, and at the same time the same node receives a
ticker update request, the update causes the node to be scheduled for
execution a second time. This has the consequence of event_counter going
out of sync, and BLE connection is ultimately lost.

The solution is to mark collsions and re-scheduled nodes as expired, and
use the increased lazy count to identify its state instead of req/ack.

Signed-off-by: Morten Priess <mtpr@oticon.com>

mtpr-ot committed:
Bluetooth: controller: Fix ticker collision update race

In the case where a BLE connection ticker node collides and does not
trigger a callback to ULL, and at the same time the same node receives a
ticker update request, the update causes the node to be scheduled for
execution a second time. This has the consequence of event_counter going
out of sync, and BLE connection is ultimately lost.

The solution is to mark collsions and re-scheduled nodes as expired, and
use the increased lazy count to identify its state instead of req/ack.

Signed-off-by: Morten Priess <mtpr@oticon.com>

mtpr-ot commented:
@cvinayak, this is my proposal for fixing the issue I mentioned on email. All our tests pass locally with 2.2.1 branch, but on master, the **advx** test is not happy. Does the error message make sense to you? It seems to say the test passed, but result is failed anyway.

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -1405,8 +1406,11 @@ static inline void ticker_job_worker_bh(struct ticker_instance *instance,
+		 * was either a collision or re-scheduled. This node should
+		 * not define the active slot reservation (slot_previous).
+		 */
+		skip_collision = ticker->lazy_current != 0U;
```
With the addition of `ticker->ack--` in worker, the request state of a collided node is now 2, and since this is not a "must_expire" case, the _slot_previous will be updated unless guarded. It must be true (?) that if `lazy_current` is non-zero, the ticker did not expire in the "old" sense and went in the air, but rather just skipped further processing.

cvinayak commented:
> In the case where a BLE connection ticker node collides and does not
trigger a callback to ULL, and at the same time the same node receives a
ticker update request

I am trying to narrow down on such a scenario, is this the case when slave connection is using slave latency, and there is a race between `ticker_update` invoked at the same time as the slave connection event collides with another state/role?

mtpr-ot commented:
> > In the case where a BLE connection ticker node collides and does not
> > trigger a callback to ULL, and at the same time the same node receives a
> > ticker update request
> 
> I am trying to narrow down on such a scenario, is this the case when slave connection is using slave latency, and there is a race between `ticker_update` invoked at the same time as the slave connection event collides with another state/role?

This specific scenario is without slave latency. But yes, the slave connection collides with ADV and does not get to execute. However, it does increment it's lazy count in `ticker_worker`. Next `ticker_job` execution then sees both an update and the "pseudo-expired" slave node, and the update thinks the slave node is no expired, and put's it pending (again) immediately.

The problem is in `ticker_job_list_manage`, which doesn't take the "else" path, because state is 1 where it should be 2.

cvinayak commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -858,10 +858,11 @@ void ticker_worker(void *param)
-				 * ticker node.
+				 * ticker node. Mark it as elapsed.
 				 */
+				ticker->ack--;
```
Thinking aloud,
1. If not `must_expire`, mark ticker instance as expired without resetting the `lazy_current`. This is desired behavior, the ticker will be scheduled at next interval period, with incremented `lazy_current`.
2. If the incremented `lazy_current` is less than `lazy_periodic`, the event callback will not be called even if `must_expire` is true. This is correct.
3. If `TICKER_RESCHEDULE_PENDING(ticker)` then mark it as expired without resetting the `lazy_current`. Would this not cause incorrect `lazy_current` ? (I think only advertiser uses this which does not bother of any incorrect `lazy` value, and hence we dont notice issues upstream?)

We need to work in consideration of changes done in https://github.com/zephyrproject-rtos/zephyr/pull/23815

and attention to this change: https://github.com/zephyrproject-rtos/zephyr/commit/b263c2860e7753beb72c70eca4e5de7fc334aa23

That is, the change/fix to reset the ticker state is inline with the fact that ticker state is put into scheduling requested state down in the code in the `ticker_job_worker_bh` function.

Now getting back to `ticker->ack--;`, a decision on whether we want `ticker->lazy_current++;` under the 3 cases (1,2,3 above) will solve the incorrect `lazy_current` that I assume causes the link loss.

Case 1 and case 2, is correctly handled. Case 3, is faulty (correct me with your independent understanding).

Finally, I have not disapproval to the addition of `ticker->ack--;` as it will put the ticker state in the correct state for it to be handled in `ticker_job_worker_bh`

mtpr-ot replied:
> 3. If `TICKER_RESCHEDULE_PENDING(ticker)` then mark it as expired with the incremented `lazy_current`. Would this not cause incorrect `lazy_current` ? (I think only advertiser uses this which does not bother of any incorrect `lazy` value, and  hence we dont notice issues upstream?)

I understand your concern for case 3. However, if I understand your point correctly, I believe this is handled in line 1828 of `ticker_job_reschedule_in_window`, where `lazy_current` is always decremented for re-scheduled nodes.

Also, I have a comprehensive ztest-suite which indicates correct functionality of the `slot_window` feature. This of course could be flawed, so please elaborate on your concern. 

I have taken the recent changes you have mentioned into consideration (and testing), and as far as I can tell, they all work well together, event if my change partly reverts some of the logic in #23815.

cvinayak replied:
Obviously, I missed the `lazy_current--` here https://github.com/zephyrproject-rtos/zephyr/blame/master/subsys/bluetooth/controller/ticker/ticker.c#L1824

Please add a similar comment in `ticker_worker` so that I dont misunderstand again in the future.

cvinayak reviewed (changes requested):
Requesting changes, to conclude on discussion, otherwise changes are acceptable with the exception of noted facts in my comments related to handling of `TICKER_RESCHEDULE_PENDING`.

Good work.

cvinayak commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -858,10 +858,11 @@ void ticker_worker(void *param)
 			if ((ticker->must_expire == 0U) ||
 			    (ticker->lazy_periodic >= ticker->lazy_current) ||
 			    TICKER_RESCHEDULE_PENDING(ticker)) {
```
```suggestion
			/* ticker->lazy_current has been decremented in the
			 * case of TICKER_RESCHEDULE_PENDING(ticker),
			 * the increment done above is cancelled and the
			 * rescheduling of the ticker will not have incorrect
			 * lazy_current.
			 */
			if ((ticker->must_expire == 0U) ||
			    (ticker->lazy_periodic >= ticker->lazy_current) ||
			    TICKER_RESCHEDULE_PENDING(ticker)) {
```

cvinayak commented:
I will trying looking into the bsim_bt test failure this week.

cvinayak reviewed (changes requested):
I think I close to the problem, it seems to be related to handling collision with single-shot tickers. Will come up with a fix and post it here. (until then requesting a change)

cvinayak committed:
Bluetooth: controller: Fix skipped single-shot ticker failure callback

A single-shot ticker in requested state beyond the elapsed
duration or skipped due to collision shall generate an
operation callback with failure status.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@mtpr-ot Feel to clean up my commit, may be have a inline functions for handling ticker state, but can be a different PR.

cvinayak reviewed (approved):
CI passed with the fix for single-shot ticker.

mtpr-ot commented:
> @mtpr-ot Feel free to clean up my commit, may be have inline functions for handling ticker state, but can be a different PR.

I think it's fine - good catch. Thanks for your help.
